effectful/ops/semantics.py

```diff
@@ -294,7 +296,24 @@ def typeof(term: Expr[T]) -> type[T]:
     from effectful.internals.runtime import interpreter
 
     with interpreter({apply: lambda _, op, *a, **k: op.__type_rule__(*a, **k)}):
-        return evaluate(term) if isinstance(term, Term) else type(term)  # type: ignore
+        if isinstance(term, Term):
+            # If term is a Term, we evaluate it to get its type
+            tp = evaluate(term)
+            if isinstance(tp, typing.TypeVar):
+                tp = (
+                    tp.__bound__
+                    if tp.__bound__
+                    else tp.__constraints__[0]
+                    if tp.__constraints__
+                    else object
+                )
+            if isinstance(tp, types.UnionType):
+                raise TypeError(
+                    f"Cannot determine type of {term} because it is a union type: {tp}"
+                )
+            return typing.get_origin(tp) or tp  # type: ignore
```

Comment:
These changes to `typeof` are necessary to preserve its current behavior under the change to `__type_rule__`. The call to `evaluate` above applies the new unification-based `__type_rule__` and may return a generic type, but we want `typeof` to continue to behave like Python's `type` which does not know anything about type variables or generic types.